### PR TITLE
fix: honor .gitignore in consultative workspace copy

### LINKED ISSUE #980: Consultative workspace copies the entire working directory per seat — 7 seats × repo size can exhaust $TMPDIR
### Summary

`_octopus_prepare_consultative_workspace` copies the **entire** working directory into a per-seat temp workspace with no exclusion of any kind. On a repository that contains large build/vendor trees, a 7-seat council needs 7× the full repository size in `$TMPDIR`, which on a default tmpfs is enough to abort the run.

In our case a council run died at the third attempt: the repo is 5.7 GB, of which 4.7 GB is vendored code that no seat ever reads, against a 7.8 GB tmpfs.

### Version

- octo **10.0.0** (also present identically in 9.65.0)
- Claude Code 2.1.251
- Linux, `/tmp` on tmpfs (7.8 GB)

### Where

`scripts/lib/agent-sync.sh`

```bash
# :139
_octopus_prepare_consultative_workspace() {
    local source_root="$1"
    local temp_root workspace
    temp_root="$(mktemp -d "${TMPDIR:-/tmp}/octopus-consultative.XXXXXX")" || return 1
    workspace="${temp_root}/workspace"
    mkdir -p "$workspace" || { rm -rf "$temp_root"; return 1; }

    if ! cp -a --reflink=auto "${source_root}/." "${workspace}/" 2>/dev/null; then
        rm -rf "$workspace"
        mkdir -p "$workspace" || { rm -rf "$temp_root"; return 1; }
        cp -a "${source_root}/." "${workspace}/" || { rm -rf "$temp_root"; return 1; }
    fi
    ...
```

and the caller, at `:176`:

```bash
source_root_logical="$PWD"
source_root="$(pwd -P)"
workspace="$(_octopus_prepare_consultative_workspace "$source_root")" || {
```

So `source_root` is simply the directory the council is launched from, copied whole.

### Measurement from our repository

| path | size | read by any seat? |
| --- | --- | --- |
| `addons/` | 3.0 GB | no |
| `apps/` | 1.7 GB | no |
| everything the council prompt referenced | ~48 MB | yes |

`addons/` and `apps/` are in `.gitignore` — they are separate repositories checked out inside the monorepo. So the data needed to skip them is already declared, and already sitting in the working tree.

### Effect

- 7 seats × 5.7 GB against a 7.8 GB tmpfs: the run aborts partway.
- The failure is not obviously about disk from the caller's side.
- Runner processes survived the aborted run and kept creating 2.3 GB workspaces, so `$TMPDIR` kept filling after the council was gone. v10's cancellation/recovery work may already address that half; the copy itself is unchanged from 9.65.0 to 10.0.0.

### Workaround we used

Launching the council from a purpose-built lean directory containing only what the prompt referenced: 48 MB instead of 5.7 GB, 120× smaller. This works because `source_root` is `$PWD`, but it means maintaining a copy that drifts from the real repository.

### Possible directions

Listed roughly by cost, no preference implied:

1. **Honour `.gitignore`** when the source root is a git work tree — `git ls-files -z | cpio` or `git archive` copies exactly the tracked set, which is also the set a reviewer would read.
2. **An exclusion env var**, e.g. `OCTOPUS_WORKSPACE_EXCLUDE`, passed through to `cp`/`rsync`.
3. **Fail early with a clear message** when the estimated size × seat count exceeds available space on `$TMPDIR`, rather than aborting mid-run.
4. **Document that `$TMPDIR` should point at disk** for large repositories — cheap, and it turns a hard failure into a slow one.

(1) alone would have avoided this entirely for us, since the 4.7 GB is already gitignored.

### Happy to help

I can test a patch against this repository — the reproduction is a single council run from a directory with a large gitignored subtree.

### PR DESCRIPTION
**Pull request by nyldn**
## Description
`_octopus_prepare_consultative_workspace` in `scripts/lib/agent-sync.sh:139` copies the source directory with `cp -a --reflink=auto "${source_root}/." "${workspace}/"` for every advisory (council/consultative) seat, with no exclusion. On a repo with large gitignored vendor/build trees, an N-seat council needs N times the full repository size in `$TMPDIR`, which can exhaust a tmpfs (reported: 7 seats x 5.7 GB against a 7.8 GB tmpfs, where 4.7 GB was gitignored and never read by any seat).

Root cause: `source_root` (`scripts/lib/agent-sync.sh:176`, `$PWD` at call time) is copied whole via `cp -a`, which has no concept of `.gitignore`.

## Fix
Adds `_octopus_copy_git_tracked_tree` (`scripts/lib/agent-sync.sh`), which, when `source_root` is inside a git work tree, pipes `git ls-files -z --cached --others --exclude-standard` through `tar` to populate the workspace with only the tracked plus untracked-but-not-ignored files — the same set a human reviewer would see, and the set option (1) in the issue asked for. If `source_root` is not a git work tree, or the git/tar path fails for any reason, `_octopus_prepare_consultative_workspace` falls back unchanged to the original full `cp -a --reflink=auto` copy, so non-git callers are unaffected.

This is the smallest fix that removes the root cause (copying gitignored bytes nobody reads); it does not add a size-estimate preflight or an exclusion env var, which the issue listed as costlier alternatives.

## Type of Change
- [x] Bug fix

## Checklist
- [x] Code passes `bash -n scripts/orchestrate.sh`
- [x] Shell scripts pass `bash -n` syntax check (ran `bash -n scripts/*.sh scripts/lib/*.sh`)
- [x] Tests pass: `bash tests/unit/test-openclaw-compat.sh` (33/33)
- [ ] OpenClaw registry in sync — not applicable, no command/skill surface changed
- [ ] New skills/commands registered — not applicable
- [ ] Version bump — not applicable, patch-level fix only
- [ ] Documentation updated — not applicable, internal implementation detail
- [ ] CHANGELOG.md updated — left for the release step if/when this merges

## Testing
- New regression test: `tests/unit/test-consultative-workspace-gitignore.sh` — builds a git work tree with a tracked file, a gitignored file, and an untracked-but-not-ignored file, and asserts the workspace produced by `_octopus_prepare_consultative_workspace` contains the first two and excludes the gitignored one; also asserts a non-git source root still gets a full copy. 2/2 pass.
- Existing `tests/unit/test-consultative-agent-dispatch.sh` still passes (8/8) — it stubs this function directly so it does not exercise the new code path, but confirms the surrounding dispatch/cleanup contract is untouched.
- `bash -n` over every script in `scripts/` and `scripts/lib/` passes.
- The full `make test-unit` (292 files) was still executing in the sandboxed session environment at push time (each test spawns subshells and git operations, which is slow there); given the change is narrowly scoped to one function with direct test coverage above, I did not block the push on it. CI will run the full suite on this PR — I'll follow up on any failures there.

## Related Issues
Closes #980

---
_Generated by [Claude Code](https://claude.ai/code/session_01REbbzLaSUtYDV1cunhrp11)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Consultative workspaces now exclude Git-ignored content while preserving tracked and other non-ignored files.
  * Nested Git work trees follow their own ignore rules.
  * Git discovery, validation, copy, and symlink safety failures now stop processing instead of falling back to a full copy.
  * Non-Git workspaces continue to use full copies.
  * Repository metadata and environment settings remain excluded from consultative sessions.
  * Interruptions and cancellation now clean up temporary workspace data reliably.

* **Tests**
  * Added coverage for ignored content, nested repositories, symlinks, cleanup, cancellation, and failure handling.

### CODE REVIEW

**Comment by nyldn** [comment]
Good catch, thanks. Verified: `git ls-files` reports a submodule (or an untracked nested checkout that isn't itself gitignored) as a single opaque path, so the tar copy was pulling that path in wholesale and bypassing its own `.gitignore` one level down.

Fixed in 7b22643d — `_octopus_copy_git_tracked_tree` now detects any such nested git work tree in the file list and re-copies it recursively with the same tracked-plus-untracked-not-ignored rule. Added a regression test (`a nested git work tree's own .gitignore is honored too`) covering it; all consultative-workspace tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
Both verified and fixed in b3c8877b:

1. **Path traversal (line 170)**: confirmed — without a symlink guard, a tracked symlink pointing outside `source_root` gets deleted and replaced with real copied files from the link target (tested this concretely: with the guard removed, `evil-link` became a real directory containing the outside repo's tracked content; with the guard, it stays a symlink). Added `! -L` to the nested-work-tree check.
2. **Failure swallowed (line 175)**: confirmed — `_octopus_copy_git_tracked_tree` runs inside the caller's `if fn; then`, which suppresses errexit for the whole call, so a bare `nested_copy || fallback` in the loop can't rely on errexit to surface a double failure. Now tracks failures in a flag and returns non-zero if any nested entry couldn't be copied either way, so `_octopus_prepare_consultative_workspace` correctly falls back to a full copy.

Both covered by new cases in `tests/unit/test-consultative-workspace-gitignore.sh`; all 5 cases in that file pass, plus the existing `test-consultative-agent-dispatch.sh` suite (8/8) and a full `bash -n` syntax check.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
`Unit Tests (macos-latest)` failed on `b3c8877b`, but not from this PR: the failing suite is `unit/test-post-tool-dispatch.sh` (the PostToolUse hook dispatcher), which this diff doesn't touch — this PR only changes `scripts/lib/agent-sync.sh` and its own test file. The failure itself is a broken-pipe error from an `echo` in that test's payload builder (`tests/unit/test-post-tool-dispatch.sh:22: echo: write error: Broken pipe`), consistent with a pipe-timing flake rather than a real regression. `Unit Tests (ubuntu-latest)` passed cleanly (292/292).

Re-ran the failed job once (`rerun_failed_jobs` on run 33293097741) to confirm.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nyldn** [comment]
@coderabbitai review

**Comment by nyldn** [comment]
Maintainer update: the rebased branch remains blocked by a TOCTOU confinement gap in the selective workspace copy. The path list is validated, then tar resolves it again; a directory swapped to an external symlink between those steps is copied successfully. I reproduced this on the exact rebased head: the function returned success and copied the external fixture bytes. Another pre- or post-copy validation would not close a swap-and-restore race. This needs an immutable or descriptor-anchored snapshot before it is safe to merge. I am leaving the PR open so issue #980 and the validated test coverage remain attached to the repair.

**Comment by nyldn** [comment]
Follow-up after exact-head spec review: three additional blockers are verified. First, Bash defers the TERM trap while the synchronous copy or provider command substitution is running; a signal sent after 0.2 seconds did not exit until the 3-second child returned. Second, repository environment clearing returns success when a readonly GIT_DIR cannot be unset, leaving the source repository reachable. Third, the non-Git cp -a path preserves absolute external symlinks, so a relative write inside the disposable workspace can mutate the original external target. Together with the archive TOCTOU race above, these require a confinement and lifecycle redesign before merge. The rebased worktree is clean and preserved; no vulnerable head was pushed.

File: tests/unit/test-consultative-workspace-gitignore.sh [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+while [[ "$#" -gt 0 ]]; do
+    case "$1" in
+        -C) archive_root="${2:-}"; shift ;;
+        -T|--files-from) list_file="${2:-}"; shift ;;
+        -T*) list_file="${1#-T}" ;;
+        --files-from=*) list_file="${1#--files-from=}" ;;
+    esac
+    shift
+done
+if [[ "$archive_root" == "$NESTED_ROOT" && -n "$list_file" && "$list_file" != "-" ]]; then
```
[the remark this answers is not in the document]
Comment (nyldn) [reply tests/unit/test-consultative-workspace-gitignore.sh]: Resolved by the current test structure. The earlier tar guard was removed; the replacement runtime case exercises the nested-copy path directly. The focused workspace suite passes 49/49 on this head.

File: tests/unit/test-consultative-workspace-gitignore.sh [hunk 2]
```diff
@@ -0,0 +1,2012 @@
...
+    plain_escape_rc=0
+else
+    plain_escape_rc=$?
+fi
+if [[ "$plain_escape_rc" -ne 0 && -z "$workspace" ]]; then
+    test_pass
+else
+    test_fail "expected an external symlink in a non-Git source to fail closed: rc=$plain_escape_rc workspace=${workspace:-none}"
+fi
+[[ -z "$workspace" ]] || command rm -rf "$(dirname "$workspace")"
```
[the remark this answers is not in the document]
Comment (nyldn) [reply tests/unit/test-consultative-workspace-gitignore.sh:653]: No change. The non-Git case already exercises the shared fail-closed symlink validator with an escaping relative target. That helper rejects every absolute target at its first branch, and the suite also covers an absolute tracked target. Another near-identical test would duplicate coverage without changing behavior. The focused workspace suite passes 49/49.
